src/simd.rs

```diff
@@ -0,0 +1,198 @@
+/// number of bytes in callback return value otherwise.
+/// This method returns final index and a possible break value.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+pub(crate) fn iterate_special_bytes<F, T>(bytes: &[u8], mut ix: usize, mut callback: F) -> (usize, Option<T>)
+    where F: FnMut(usize) -> LoopInstruction<Option<T>> 
+{
+    let upperbound = bytes.len().saturating_sub(15);
+    
+    while ix < upperbound {
+        let mut mask = unsafe { compute_mask(bytes, ix as isize) };
+        let block_start = ix;
+
+        while mask != 0 {
+            let mask_ix = mask.trailing_zeros() as usize;
+            ix += mask_ix;
+            match callback(ix) {
+                LoopInstruction::ContinueAndSkip(skip) => {
+                    ix += skip + 1;
+                    mask >>= skip + mask_ix + 1;
+                }
+                LoopInstruction::BreakAtWith(ix, val) => return (ix, val),
+            }
+        }
+
+        ix = std::cmp::max(ix, 16 + block_start);
+    }
+
+    let alignment = (bytes.as_ptr() as usize) % 16;
+    let remaining_bytes = bytes.len().saturating_sub(ix);
+
+    if remaining_bytes > 0 {
+        let mut mask = if alignment + remaining_bytes > 16 {
+            unsafe { compute_mask(bytes, ix as isize) }
+        } else {
+            // shift off the bytes we aren't interested in
+            unsafe { compute_mask(bytes, ix as isize - alignment as isize) >> alignment } 
+        };
+
+        // zero out the bytes we read past end of buffer
+        mask &= (1 << remaining_bytes) - 1;
+
+        while mask != 0 {
+            let mask_ix = mask.trailing_zeros() as usize;
+            ix += mask_ix;
+            match callback(ix) {
+                LoopInstruction::ContinueAndSkip(skip) => {
+                    ix += skip + 1;
+                    mask >>= skip + 1 + mask_ix;
+                }
+                LoopInstruction::BreakAtWith(ix, val) => return (ix, val),
+            }
+        }
+    }
+
+    (bytes.len(), None)
+}
+
+/// Scalar fallback.
+#[cfg(not(all(target_arch = "x86_64", feature="simd")))]
+pub(crate) fn iterate_special_bytes<F, T>(bytes: &[u8], mut ix: usize, mut callback: F) -> (usize, Option<T>)
+    where F: FnMut(usize) -> LoopInstruction<Option<T>> 
+{
+    while ix < bytes.len() {
+        match callback(ix) {
+            LoopInstruction::ContinueAndSkip(skip) => {
+                ix += skip + 1;
+            }
+            LoopInstruction::BreakAtWith(ix, val) => {
+                return (ix, val);
+            }
+        }
+    }
+
+    (ix, None)
+}
+
+#[cfg(all(target_arch = "x86_64", feature="simd", test))]
+mod simd_test {
+    use super::{iterate_special_bytes, LoopInstruction};
+
+    fn check_expected_indices(bytes: &[u8], expected: &[usize], skip: usize) {
+        let mut indices = vec![];
+
+        iterate_special_bytes(bytes, 0, |ix| {
+            indices.push(ix);
+
+            // for type inference
```

Comment:
I wouldn't be surprised if there were a better way to pin down the type, but ok.

Reply:
There is, not sure why I didn't just typehint here. Will fix.